### PR TITLE
Reward distribution mechanics

### PR DESCRIPTION
**Pull request by michaelfig**
refs: #4399

## Description

Implement 10% of reward pool distributed at the start of every 24h epoch.

### CODE REVIEW

**Review by arirubinstein — APPROVED** [review]

File: golang/cosmos/app/ante/fees.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// If the first signer does not have the funds to pay for the fees, return with InsufficientFunds error
+// Call next AnteHandler if fees successfully deducted
+// CONTRACT: Tx must implement FeeTx interface to use DeductFeeDecorator
+type DeductFeeDecorator struct {
+	ak               AccountKeeper
+	bankKeeper       types.BankKeeper
+	feegrantKeeper   FeegrantKeeper
+	feeCollectorName string
+}
+
```
Comment (JimLarson) [inline golang/cosmos/app/ante/fees.go:21]: Add the check that it's an `AnteDecorator`:

```
var _ sdk.AnteDecorator = DeductFeeDecorator{}
```

File: golang/cosmos/x/vbank/keeper/rewards.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import (
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// minCoins returns the minimum of each denomination.
+// The input coins should be sorted.
+func minCoins(a, b sdk.Coins) sdk.Coins {
```
Comment (JimLarson) [inline golang/cosmos/x/vbank/keeper/rewards.go:11]: This function is now `sdk.Coins.Min()`.

File: golang/cosmos/proto/agoric/vbank/vbank.proto [hunk 3]
```diff
@@ -11,11 +11,27 @@ message Params {
...
-    // the full fee buffer should be distributed immediately.
-    int64 fee_epoch_duration_blocks = 1 [
-      (gogoproto.moretags) = "yaml:\"fee_epoch_duration_blocks\""
+    // the full reward buffer should be distributed immediately.
+    int64 reward_epoch_duration_blocks = 1 [
+      (gogoproto.moretags) = "yaml:\"reward_epoch_duration_blocks\""
+    ];
+
+    // per_epoch_reward_fraction is a fraction of the reward pool to distrubute
+    // once every reward epoch.  If less than zero, use approximately continuous
```
Comment (JimLarson) [inline golang/cosmos/proto/agoric/vbank/vbank.proto:22]: The less-than-zero behavior doesn't seem to be implemented and parameter validation disallows it.

File: golang/cosmos/x/vbank/vbank_test.go [hunk 4]
```diff
@@ -398,11 +398,28 @@ func Test_Receive_GiveToFeeCollector(t *testing.T) {
 				sdk.NewCoin("yoctoquatloos", sdk.NewInt(123456789123456789).MulRaw(1000).AddRaw(124)),
 			),
 		},
+		{
+			name:          "big",
+			duration:      1000 * 1000,
+			rewardPool:    sdk.NewCoins(),
+			feeAmount:     "123456789123456789123456789",
+			feeDenom:      "yoctoquatloos",
```
Comment (JimLarson) [inline golang/cosmos/x/vbank/vbank_test.go:406]: You don't have much faith in the newcomer.

**Review by JimLarson — COMMENTED** [review]
After-the-fact review with a few things that should eventually be cleaned up.

**Pull request merged**